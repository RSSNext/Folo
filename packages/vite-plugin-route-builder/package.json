{
  "name": "@follow-app/vite-plugin-route-builder",
  "type": "module",
  "version": "0.0.1",
  "exports": {
    ".": "./src/index.ts"
  },
  "main": "src/index.ts",
  "types": "src/index.ts",
  "files": [
    "src"
  ],
  "scripts": {
    "build": "tsdown",
    "test": "vitest",
    "test:run": "vitest run"
  },
  "devDependencies": {
    "@types/node": "22.15.23",
    "es-toolkit": "1.38.0",
    "fast-glob": "3.3.3",
    "happy-dom": "17.5.6",
    "react-router": "7.6.1",
    "tsdown": "0.12.8",
    "typescript": "catalog:",
    "vite": "6.3.5",
    "vite-tsconfig-paths": "5.1.4",
    "vitest": "3.1.4"
  },
  "publishConfig": {
    "access": "public",
    "main": "dist/index.js",
    "types": "dist/index.d.ts"
  }
}
