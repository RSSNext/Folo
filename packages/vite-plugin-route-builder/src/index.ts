export { default } from "./vite-plugin-route-builder"
