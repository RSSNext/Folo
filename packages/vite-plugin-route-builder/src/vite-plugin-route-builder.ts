import { writeFileSync } from "node:fs"
import { dirname, relative, resolve } from "node:path"
import { inspect } from "node:util"

import glob from "fast-glob"
import type { RouteObject } from "react-router"
import type { Logger, Plugin } from "vite"

import { buildGlobRoutes } from "./utils/route-builder"

export interface RouteBuilderPluginOptions {
  /** Page files glob pattern */
  pagePattern?: string
  /** Output path for generated routes */
  outputPath?: string
  /** Whether to enable in dev mode */
  enableInDev?: boolean
  /** Custom file to route path transformation logic */
  transformPath?: (path: string) => string
}

export function routeBuilderPluginV2(options: RouteBuilderPluginOptions = {}): Plugin {
  const {
    pagePattern = "./pages/**/*.tsx",
    outputPath = "./src/generated-routes.ts",
    enableInDev = true,
    transformPath,
  } = options

  let isProduction = false
  let root = ""
  let logger: Logger

  function generateRouteFileContent(
    routes: RouteObject[],
    fileToImportMap: Record<string, string>,
  ): string {
    // Collect all used lazy functions
    const usedLazyFunctions = new Set<string>()
    const lazyFunctionMap = new Map<string, string>()
    let lazyCounter = 0

    // Recursively traverse route tree, collect all used lazy functions
    function collectUsedLazyFunctions(routes: RouteObject[]) {
      routes.forEach((route) => {
        if (route.lazy && route.handle?.fs) {
          const fsPath = route.handle.fs

          // Try to find the corresponding file
          let matchedKey: string | undefined

          // Direct match strategy: according to route-builder.ts logic
          // 1. For grouped routes, fs is segmentPathKey, need to find ${fs}/layout.tsx
          // 2. For layout files, fs is segmentPathKey, need to find ${fs}.tsx
          // 3. For normal pages, fs is ${segmentPathKey}/${normalizeKey}, but the actual file is ${segmentPathKey}.tsx

          // Strategy 1: Direct match fs.tsx
          if (fileToImportMap[`${fsPath}.tsx`]) {
            matchedKey = `${fsPath}.tsx`
          }
          // Strategy 2: layout file (for grouped routes)
          else if (fileToImportMap[`${fsPath}/layout.tsx`]) {
            matchedKey = `${fsPath}/layout.tsx`
          }
          // Strategy 3: index file
          else if (fileToImportMap[`${fsPath}/index.tsx`]) {
            matchedKey = `${fsPath}/index.tsx`
          }
          // Strategy 4: For special path correction
          else {
            // If fsPath ends with /, it might be an index page
            if (fsPath.endsWith("/")) {
              const correctedPath = fsPath.slice(0, -1) // Remove trailing /
              if (fileToImportMap[`${correctedPath}/index.tsx`]) {
                matchedKey = `${correctedPath}/index.tsx`
              } else if (fileToImportMap[`${correctedPath}.tsx`]) {
                matchedKey = `${correctedPath}.tsx`
              }
            }
            // For dynamic routes, remove /:param part, keep file path
            else if (fsPath.includes("/:")) {
              // Handle cases like "./pages/category/[category]/:category"
              // Remove /:param part, keep the previous path
              const correctedPath = fsPath.replace(/\/:[^/]+(?:\/.*)?$/, "")
              if (fileToImportMap[`${correctedPath}.tsx`]) {
                matchedKey = `${correctedPath}.tsx`
              }
            }
            // If fsPath ends with a repeated path segment, remove the last segment
            else {
              const pathParts = fsPath.split("/")
              if (pathParts.length >= 2) {
                const lastPart = pathParts.at(-1)
                const secondLastPart = pathParts.at(-2)

                // If the last two path segments are the same, remove the last one
                if (lastPart === secondLastPart) {
                  const correctedPath = pathParts.slice(0, -1).join("/")
                  if (fileToImportMap[`${correctedPath}.tsx`]) {
                    matchedKey = `${correctedPath}.tsx`
                  }
                }
              }
            }
          }

          if (matchedKey && fileToImportMap[matchedKey]) {
            const lazyFuncName = `lazy${lazyCounter++}`
            usedLazyFunctions.add(matchedKey)
            lazyFunctionMap.set(matchedKey, lazyFuncName)

            logger.info(
              `[route-builder-v2] Mapped lazy function: ${fsPath} -> ${matchedKey} -> ${lazyFuncName}`,
            )
          } else {
            logger.warn(`[route-builder-v2] Could not find file for fs path: ${fsPath}`)
            logger.warn(
              `[route-builder-v2] Available file keys: ${inspect(Object.keys(fileToImportMap), {
                depth: null,
              })}`,
            )
          }
        }

        if (route.children) {
          collectUsedLazyFunctions(route.children)
        }
      })
    }

    collectUsedLazyFunctions(routes)

    // Generate import statements
    const imports: string[] = []
    usedLazyFunctions.forEach((key) => {
      const importPath = fileToImportMap[key]
      const lazyFuncName = lazyFunctionMap.get(key)
      if (importPath && lazyFuncName) {
        imports.push(`const ${lazyFuncName} = () => import("${importPath}")`)
      }
    })

    // Recursively process routes, replace lazy functions and remove handle
    function processRoutes(routes: RouteObject[]): any {
      return routes.map((route) => {
        const newRoute: any = { ...route }

        // Process lazy functions
        if (route.lazy && route.handle?.fs) {
          const fsPath = route.handle.fs

          // Find matching file - use the same matching logic
          let matchedKey: string | undefined

          if (fileToImportMap[`${fsPath}.tsx`]) {
            matchedKey = `${fsPath}.tsx`
          } else if (fileToImportMap[`${fsPath}/layout.tsx`]) {
            matchedKey = `${fsPath}/layout.tsx`
          } else if (fileToImportMap[`${fsPath}/index.tsx`]) {
            matchedKey = `${fsPath}/index.tsx`
          } else {
            // For special path correction
            if (fsPath.endsWith("/")) {
              const correctedPath = fsPath.slice(0, -1) // Remove trailing /
              if (fileToImportMap[`${correctedPath}/index.tsx`]) {
                matchedKey = `${correctedPath}/index.tsx`
              } else if (fileToImportMap[`${correctedPath}.tsx`]) {
                matchedKey = `${correctedPath}.tsx`
              }
            }
            // For dynamic routes, remove /:param part, keep file path
            else if (fsPath.includes("/:")) {
              // Handle cases like "./pages/category/[category]/:category"
              // Remove /:param part, keep the previous path
              const correctedPath = fsPath.replace(/\/:[^/]+(?:\/.*)?$/, "")
              if (fileToImportMap[`${correctedPath}.tsx`]) {
                matchedKey = `${correctedPath}.tsx`
              }
            }
            // For repeated path segments, try removing the last segment
            else {
              const pathParts = fsPath.split("/")
              if (pathParts.length >= 2) {
                const lastPart = pathParts.at(-1)
                const secondLastPart = pathParts.at(-2)

                // If the last two path segments are the same, remove the last one
                if (lastPart === secondLastPart) {
                  const correctedPath = pathParts.slice(0, -1).join("/")
                  if (fileToImportMap[`${correctedPath}.tsx`]) {
                    matchedKey = `${correctedPath}.tsx`
                  }
                }
              }
            }
          }

          if (matchedKey && lazyFunctionMap.has(matchedKey)) {
            newRoute.lazy = `__LAZY_${lazyFunctionMap.get(matchedKey)}__`
          } else {
            // If no matching file is found, delete lazy property
            delete newRoute.lazy
            logger.warn(`[route-builder-v2] No lazy function for route: ${fsPath}`)
          }
        }

        // Remove handle property
        delete newRoute.handle

        // Recursively process children
        if (route.children) {
          newRoute.children = processRoutes(route.children)
        }

        return newRoute
      })
    }

    const processedRoutes = processRoutes(routes)

    // Convert routes object to string and replace lazy function placeholders
    const routesString = JSON.stringify(processedRoutes, null, 2).replaceAll(
      /"__LAZY_(\w+)__"/g,
      "$1",
    )

    return `// This file is auto-generated by vite-plugin-route-builder
// Do not edit manually
/* eslint-disable */
// @ts-nocheck

import type { RouteObject } from "react-router"

// Lazy imports for page components
${imports.join("\n")}

// Generated route configuration
export const routes: RouteObject[] = ${routesString}

export default routes
`
  }

  function generateRoutes() {
    try {
      const pageFiles = glob.sync(pagePattern, {
        cwd: root,
        absolute: true,
      })

      logger.info(`[route-builder-v2] Found ${pageFiles.length} page files`)

      // Build glob object, key is the relative path to pages directory
      const globObject: Record<string, () => Promise<any>> = {}
      const fileToImportMap: Record<string, string> = {}

      pageFiles.forEach((absolutePath) => {
        // Get relative path to root
        const relativePath = relative(root, absolutePath)

        // Convert to ./pages/ format for route-builder
        let routeKey: string
        if (relativePath.includes("/pages/")) {
          routeKey = `./pages/${relativePath.split("/pages/")[1]}`
        } else if (relativePath.includes("\\pages\\")) {
          routeKey = `./pages/${relativePath.split("\\pages\\")[1]?.replaceAll("\\", "/")}`
        } else {
          // Assume file is in pages directory
          routeKey = `./${relativePath.replaceAll("\\", "/")}`
        }

        // Apply custom path transformation
        if (transformPath) {
          routeKey = transformPath(routeKey)
        }

        // Generate relative path for import (relative to output file)
        const outputDir = dirname(resolve(root, outputPath))
        let importPath = relative(outputDir, absolutePath)

        // Ensure correct path separator
        importPath = importPath.replaceAll("\\", "/")

        // Ensure relative path starts with ./ or ../
        if (!importPath.startsWith(".")) {
          importPath = `./${importPath}`
        }

        // Remove .tsx extension
        importPath = importPath.replace(/\.tsx$/, "")

        globObject[routeKey] = () => Promise.resolve({ default: () => null })
        fileToImportMap[routeKey] = importPath

        logger.info(`[route-builder-v2] Mapped: ${routeKey} -> ${importPath}`)
      })

      // Use existing route building logic
      const routes = buildGlobRoutes(globObject)

      // Generate route file content
      const routeFileContent = generateRouteFileContent(routes, fileToImportMap)

      const outputFilePath = resolve(root, outputPath)
      writeFileSync(outputFilePath, routeFileContent, "utf-8")

      logger.info(`[route-builder-v2] Generated routes: ${outputFilePath}`)
    } catch (error: any) {
      logger.error(`[route-builder-v2] Error generating routes:${error.message}`)
      console.error(error)
      throw error
    }
  }

  return {
    name: "vite-plugin-route-builder-v2",
    configResolved(config) {
      isProduction = config.command === "build"
      root = config.root
      logger = config.logger
    },

    buildStart() {
      if (isProduction || enableInDev) {
        generateRoutes()
      }
    },

    configureServer(server) {
      if (!enableInDev) return

      const watchPattern = resolve(root, pagePattern.replace("./", ""))
      server.watcher.add(watchPattern)

      server.watcher.on("add", handleFileChange)
      server.watcher.on("unlink", handleFileChange)

      function handleFileChange(path: string) {
        const relativePath = relative(root, path)
        if (relativePath.includes("/pages/") && relativePath.endsWith(".tsx")) {
          logger.info(`[route-builder-v2] Page file changed: ${relativePath}`)
          generateRoutes()

          // Send custom HMR event
          server.ws.send({
            type: "custom",
            event: "routes-updated",
            data: { timestamp: Date.now() },
          })
        }
      }
    },
  }
}

export default routeBuilderPluginV2
