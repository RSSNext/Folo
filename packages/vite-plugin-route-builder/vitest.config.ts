import tsconfigPath from "vite-tsconfig-paths"
import { defineConfig } from "vitest/config"

export default defineConfig({
  test: {
    globals: true,
    environment: "node",
    coverage: {
      reporter: ["text", "json", "html"],
      exclude: ["node_modules/", "src/__tests__/", "dist/", "*.config.*", "**/*.d.ts"],
    },
  },
  plugins: [
    tsconfigPath({
      projects: ["./tsconfig.json"],
    }),
  ],
})
