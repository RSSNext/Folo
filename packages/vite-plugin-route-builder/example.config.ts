import { defineConfig } from "vite"

import { routeBuilderPluginV2 } from "./src/vite-plugin-route-builder"

export default defineConfig({
  plugins: [
    routeBuilderPluginV2({
      // Glob pattern for page files
      pagePattern: "./src/pages/**/*.{tsx,sync.tsx}",

      // Output path for generated routes
      outputPath: "./src/generated-routes.ts",

      // Enable in dev mode for hot reload
      enableInDev: true,

      // Debug logging
      debug: true,

      // Custom segment group ordering: main routes first, then admin, then external
      // Both formats supported: with or without parentheses
      segmentGroupOrder: ["(main)", "(admin)", "(external)"],

      // Optional: transform file paths
      transformPath: (path: string) => {
        // Example: rename admin pages
        return path.replace("/admin/", "/dashboard/")
      },
    }),
  ],
})

/* 
File structure example:

src/pages/
├── index.tsx              → / (lazy loaded)
├── about.tsx              → /about (lazy loaded)
├── critical.sync.tsx      → /critical (sync loaded)
├── settings/
│   ├── layout.tsx         → /settings (lazy loaded)
│   ├── index.tsx          → /settings/ (lazy loaded)
│   └── profile.sync.tsx   → /settings/profile (sync loaded)
├── (main)/                → Route group (sorted first due to segmentGroupOrder)
│   ├── layout.tsx         → / (lazy loaded)
│   └── dashboard.tsx      → /dashboard (lazy loaded)
├── (admin)/               → Route group (sorted second due to segmentGroupOrder)
│   ├── layout.tsx         → / (lazy loaded)
│   └── users.tsx          → /users (lazy loaded)
└── (external)/            → Route group (sorted third due to segmentGroupOrder)
    ├── layout.tsx         → / (lazy loaded)
    └── api-docs.tsx       → /api-docs (lazy loaded)

Generated output (route groups ordered by segmentGroupOrder):

import * as SyncComponent0 from "./pages/critical.sync"
import * as SyncComponent1 from "./pages/settings/profile.sync"
const lazy0 = () => import("./pages/index")
const lazy1 = () => import("./pages/about")
const lazy2 = () => import("./pages/(main)/layout")
const lazy3 = () => import("./pages/(admin)/layout")
const lazy4 = () => import("./pages/(external)/layout")
// ... more imports

export const routes: RouteObject[] = [
  // Non-grouped routes first
  {
    path: "",
    lazy: lazy0,
  },
  {
    path: "critical",
    Component: SyncComponent0.Component,
    loader: SyncComponent0.loader,
  },
  // Route groups ordered by segmentGroupOrder: main, admin, external
  {
    path: "",
    lazy: lazy2, // (main) group layout
    children: [] // main routes
  },
  {
    path: "",
    lazy: lazy3, // (admin) group layout  
    children: [] // admin routes
  },
  {
    path: "",
    lazy: lazy4, // (external) group layout
    children: [] // external routes
  },
  // ... more routes
]
*/
