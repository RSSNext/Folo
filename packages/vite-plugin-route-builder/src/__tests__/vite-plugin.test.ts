import * as fs from "node:fs"

import { afterEach, beforeEach, describe, expect, it, vi } from "vitest"

import { routeBuilderPluginV2 } from "../vite-plugin-route-builder"

// Mock file system and glob
vi.mock("node:fs", () => ({
  writeFileSync: vi.fn(),
  promises: {
    access: vi.fn(),
    readFile: vi.fn(),
    writeFile: vi.fn(),
  },
  resolve: vi.fn(),
  dirname: vi.fn(),
  relative: vi.fn(),
}))

vi.mock("fast-glob", () => ({
  default: {
    sync: vi.fn(),
  },
}))

vi.mock("pathe", () => ({
  resolve: vi.fn((root: string, path: string) => `${root}/${path}`),
  relative: vi.fn((from: string, to: string) => to.replace(from, "")),
  dirname: vi.fn((path: string) => path.split("/").slice(0, -1).join("/")),
}))

vi.mock("./utils/route-builder", () => ({
  buildGlobRoutes: vi.fn(() => [
    {
      path: "test",
      lazy: () => Promise.resolve({ default: () => null }),
      handle: {
        fs: "./pages/test",
        fullPath: "/test",
        isSync: false,
      },
    },
  ]),
}))

// Helper function to capture generated content
function captureGeneratedContent(): { content: string; filePath: string } | null {
  const mockWriteFileSync = vi.mocked(fs.writeFileSync)

  if (!mockWriteFileSync.mock || mockWriteFileSync.mock.calls.length === 0) {
    return null
  }

  const lastCall = mockWriteFileSync.mock.calls.at(-1)
  if (!lastCall) {
    return null
  }

  return {
    filePath: lastCall[0] as string,
    content: lastCall[1] as string,
  }
}

describe("routeBuilderPluginV2", () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  afterEach(() => {
    vi.restoreAllMocks()
  })

  it("should create plugin with default options", () => {
    const plugin = routeBuilderPluginV2()

    expect(plugin.name).toBe("vite-plugin-route-builder-v2")
    expect(plugin).toHaveProperty("configResolved")
    expect(plugin).toHaveProperty("buildStart")
    expect(plugin).toHaveProperty("configureServer")
  })

  it("should use custom options", () => {
    const customOptions = {
      pagePattern: "./custom/**/*.tsx",
      outputPath: "./custom/routes.ts",
      enableInDev: false,
      debug: true,
    }

    const plugin = routeBuilderPluginV2(customOptions)
    expect(plugin.name).toBe("vite-plugin-route-builder-v2")
  })

  it("should generate routes for sync and async files", async () => {
    const mockFiles = [
      "/project/src/pages/index.tsx",
      "/project/src/pages/about.sync.tsx",
      "/project/src/pages/settings/layout.tsx",
      "/project/src/pages/settings/profile.sync.tsx",
    ]

    const glob = await import("fast-glob")
    vi.mocked(glob.default.sync).mockReturnValue(mockFiles)

    const plugin = routeBuilderPluginV2({
      pagePattern: "./src/pages/**/*.{tsx,sync.tsx}",
      outputPath: "./src/generated-routes.ts",
    })

    const mockConfig = {
      command: "build" as const,
      root: "/project",
      logger: {
        info: vi.fn(),
        warn: vi.fn(),
        error: vi.fn(),
      },
    }

    // Simulate plugin lifecycle
    if (typeof plugin.configResolved === "function") {
      plugin.configResolved(mockConfig as any)
    }
    if (typeof plugin.buildStart === "function") {
      await plugin.buildStart.call({} as any, {} as any)
    }

    expect(glob.default.sync).toHaveBeenCalledWith("./src/pages/**/*.{tsx,sync.tsx}", {
      cwd: "/project",
      absolute: true,
    })
  })

  it("should generate correct import statements for sync files", async () => {
    const mockFiles = ["/project/src/pages/critical.sync.tsx", "/project/src/pages/normal.tsx"]

    const glob = await import("fast-glob")
    vi.mocked(glob.default.sync).mockReturnValue(mockFiles)

    const plugin = routeBuilderPluginV2({
      outputPath: "./src/generated-routes.ts",
      debug: true,
    })

    const mockConfig = {
      command: "build" as const,
      root: "/project",
      logger: {
        info: vi.fn(),
        warn: vi.fn(),
        error: vi.fn(),
      },
    }

    const { writeFileSync } = await import("node:fs")
    const mockWriteFileSync = vi.mocked(writeFileSync)

    if (typeof plugin.configResolved === "function") {
      plugin.configResolved(mockConfig as any)
    }
    if (typeof plugin.buildStart === "function") {
      await plugin.buildStart.call({} as any, {} as any)
    }

    expect(mockWriteFileSync).toHaveBeenCalled()
    expect(mockConfig.logger.info).toHaveBeenCalledWith(
      expect.stringContaining("Generated routes:"),
    )

    // Verify generated content with snapshot
    const generatedContent = captureGeneratedContent()
    expect(generatedContent).toBeTruthy()
    expect(generatedContent!.content).toMatchSnapshot("sync-and-async-routes")
  })

  it("should handle transformPath option", async () => {
    const mockFiles = ["/project/src/pages/test.tsx"]

    const glob = await import("fast-glob")
    vi.mocked(glob.default.sync).mockReturnValue(mockFiles)

    const plugin = routeBuilderPluginV2({
      transformPath: (path) => path.replace("./pages/", "./custom/"),
      debug: true,
    })

    const mockConfig = {
      command: "build" as const,
      root: "/project",
      logger: {
        info: vi.fn(),
        warn: vi.fn(),
        error: vi.fn(),
      },
    }

    if (typeof plugin.configResolved === "function") {
      plugin.configResolved(mockConfig as any)
    }
    if (typeof plugin.buildStart === "function") {
      await plugin.buildStart.call({} as any, {} as any)
    }

    expect(mockConfig.logger.info).toHaveBeenCalledWith(
      expect.stringContaining("./custom/test.tsx"),
    )
  })

  it("should watch files in dev mode", () => {
    const plugin = routeBuilderPluginV2({
      enableInDev: true,
    })

    const mockWatcher = {
      add: vi.fn(),
      on: vi.fn(),
    }

    const mockServer = {
      watcher: mockWatcher,
      ws: {
        send: vi.fn(),
      },
    }

    const mockConfig = {
      command: "serve" as const,
      root: "/project",
      logger: {
        info: vi.fn(),
        warn: vi.fn(),
        error: vi.fn(),
      },
    }

    if (typeof plugin.configResolved === "function") {
      plugin.configResolved(mockConfig as any)
    }
    if (typeof plugin.configureServer === "function") {
      plugin.configureServer(mockServer as any)
    }

    expect(mockWatcher.add).toHaveBeenCalled()
    expect(mockWatcher.on).toHaveBeenCalledWith("add", expect.any(Function))
    expect(mockWatcher.on).toHaveBeenCalledWith("unlink", expect.any(Function))
  })

  it("should not generate in dev mode when disabled", async () => {
    const plugin = routeBuilderPluginV2({
      enableInDev: false,
    })

    const mockConfig = {
      command: "serve" as const,
      root: "/project",
      logger: {
        info: vi.fn(),
        warn: vi.fn(),
        error: vi.fn(),
      },
    }

    const glob = await import("fast-glob")

    if (typeof plugin.configResolved === "function") {
      plugin.configResolved(mockConfig as any)
    }
    if (typeof plugin.buildStart === "function") {
      await plugin.buildStart.call({} as any, {} as any)
    }

    expect(glob.default.sync).not.toHaveBeenCalled()
  })

  describe("Generated Routes Snapshots", () => {
    it("should generate correct routes for mixed sync and async files", async () => {
      const mockFiles = [
        "/project/src/pages/index.tsx",
        "/project/src/pages/about.sync.tsx",
        "/project/src/pages/settings/layout.tsx",
        "/project/src/pages/settings/profile.sync.tsx",
        "/project/src/pages/blog/[id].tsx",
        "/project/src/pages/dashboard/admin.sync.tsx",
      ]

      const glob = await import("fast-glob")
      vi.mocked(glob.default.sync).mockReturnValue(mockFiles)

      const plugin = routeBuilderPluginV2({
        pagePattern: "./src/pages/**/*.{tsx,sync.tsx}",
        outputPath: "./src/generated-routes.ts",
      })

      const mockConfig = {
        command: "build" as const,
        root: "/project",
        logger: {
          info: vi.fn(),
          warn: vi.fn(),
          error: vi.fn(),
        },
      }

      if (typeof plugin.configResolved === "function") {
        plugin.configResolved(mockConfig as any)
      }
      if (typeof plugin.buildStart === "function") {
        await plugin.buildStart.call({} as any, {} as any)
      }

      const generatedContent = captureGeneratedContent()
      expect(generatedContent).toBeTruthy()
      expect(generatedContent!.content).toMatchSnapshot("mixed-routes-structure")
    })

    it("should generate correct routes for only sync files", async () => {
      const mockFiles = [
        "/project/src/pages/critical.sync.tsx",
        "/project/src/pages/important.sync.tsx",
        "/project/src/pages/settings/config.sync.tsx",
      ]

      const glob = await import("fast-glob")
      vi.mocked(glob.default.sync).mockReturnValue(mockFiles)

      const plugin = routeBuilderPluginV2({
        outputPath: "./src/generated-routes.ts",
      })

      const mockConfig = {
        command: "build" as const,
        root: "/project",
        logger: {
          info: vi.fn(),
          warn: vi.fn(),
          error: vi.fn(),
        },
      }

      if (typeof plugin.configResolved === "function") {
        plugin.configResolved(mockConfig as any)
      }
      if (typeof plugin.buildStart === "function") {
        await plugin.buildStart.call({} as any, {} as any)
      }

      const generatedContent = captureGeneratedContent()
      expect(generatedContent).toBeTruthy()
      expect(generatedContent!.content).toMatchSnapshot("sync-only-routes")
    })

    it("should generate correct routes for only async files", async () => {
      const mockFiles = [
        "/project/src/pages/index.tsx",
        "/project/src/pages/about.tsx",
        "/project/src/pages/contact.tsx",
        "/project/src/pages/blog/[slug].tsx",
      ]

      const glob = await import("fast-glob")
      vi.mocked(glob.default.sync).mockReturnValue(mockFiles)

      const plugin = routeBuilderPluginV2({
        outputPath: "./src/generated-routes.ts",
      })

      const mockConfig = {
        command: "build" as const,
        root: "/project",
        logger: {
          info: vi.fn(),
          warn: vi.fn(),
          error: vi.fn(),
        },
      }

      if (typeof plugin.configResolved === "function") {
        plugin.configResolved(mockConfig as any)
      }
      if (typeof plugin.buildStart === "function") {
        await plugin.buildStart.call({} as any, {} as any)
      }

      const generatedContent = captureGeneratedContent()
      expect(generatedContent).toBeTruthy()
      expect(generatedContent!.content).toMatchSnapshot("async-only-routes")
    })

    it("should generate correct routes with custom transformPath", async () => {
      const mockFiles = ["/project/src/pages/test.tsx", "/project/src/pages/demo.sync.tsx"]

      const glob = await import("fast-glob")
      vi.mocked(glob.default.sync).mockReturnValue(mockFiles)

      const plugin = routeBuilderPluginV2({
        transformPath: (path) => path.replace("./pages/", "./custom/"),
        outputPath: "./src/generated-routes.ts",
      })

      const mockConfig = {
        command: "build" as const,
        root: "/project",
        logger: {
          info: vi.fn(),
          warn: vi.fn(),
          error: vi.fn(),
        },
      }

      if (typeof plugin.configResolved === "function") {
        plugin.configResolved(mockConfig as any)
      }
      if (typeof plugin.buildStart === "function") {
        await plugin.buildStart.call({} as any, {} as any)
      }

      const generatedContent = captureGeneratedContent()
      expect(generatedContent).toBeTruthy()
      expect(generatedContent!.content).toMatchSnapshot("custom-transform-routes")
    })

    it("should generate empty routes array when no files found", async () => {
      const mockFiles: string[] = []

      const glob = await import("fast-glob")
      vi.mocked(glob.default.sync).mockReturnValue(mockFiles)

      const plugin = routeBuilderPluginV2({
        outputPath: "./src/generated-routes.ts",
      })

      const mockConfig = {
        command: "build" as const,
        root: "/project",
        logger: {
          info: vi.fn(),
          warn: vi.fn(),
          error: vi.fn(),
        },
      }

      if (typeof plugin.configResolved === "function") {
        plugin.configResolved(mockConfig as any)
      }
      if (typeof plugin.buildStart === "function") {
        await plugin.buildStart.call({} as any, {} as any)
      }

      const generatedContent = captureGeneratedContent()
      expect(generatedContent).toBeTruthy()
      expect(generatedContent!.content).toMatchSnapshot("empty-routes")
    })
  })
})
