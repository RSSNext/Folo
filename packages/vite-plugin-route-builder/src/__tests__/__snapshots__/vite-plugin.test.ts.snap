// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`routeBuilderPluginV2 > Generated Routes Snapshots > should generate correct routes for mixed sync and async files > mixed-routes-structure 1`] = `
"// This file is auto-generated by vite-plugin-route-builder
// Do not edit manually
/* eslint-disable */
// @ts-nocheck

import type { RouteObject } from "react-router"

// Imports for page components
import * as SyncComponent0 from ".//project/src/pages/about.sync"
import * as SyncComponent1 from ".//project/src/pages/dashboard/admin.sync"
import * as SyncComponent2 from ".//project/src/pages/settings/profile.sync"
const lazy0 = () => import(".//project/src/pages/blog/[id]")
const lazy1 = () => import(".//project/src/pages/settings/layout")
const lazy2 = () => import(".//project/src/pages/index")

// Generated route configuration
export const routes: RouteObject[] = [
  {
    "path": "about",
    "Component": SyncComponent0.Component,
    "loader": SyncComponent0.loader
  },
  {
    "path": "blog",
    "children": [
      {
        "path": ":id",
        "lazy": lazy0
      }
    ]
  },
  {
    "path": "dashboard",
    "children": [
      {
        "path": "admin",
        "Component": SyncComponent1.Component,
        "loader": SyncComponent1.loader
      }
    ]
  },
  {
    "path": "settings",
    "children": [
      {
        "path": "",
        "lazy": lazy1,
        "children": [
          {
            "path": "profile",
            "Component": SyncComponent2.Component,
            "loader": SyncComponent2.loader
          }
        ]
      }
    ]
  },
  {
    "path": "",
    "lazy": lazy2
  }
]

export default routes
"
`;

exports[`routeBuilderPluginV2 > Generated Routes Snapshots > should generate correct routes for only async files > async-only-routes 1`] = `
"// This file is auto-generated by vite-plugin-route-builder
// Do not edit manually
/* eslint-disable */
// @ts-nocheck

import type { RouteObject } from "react-router"

// Imports for page components
const lazy0 = () => import(".//project/src/pages/about")
const lazy1 = () => import(".//project/src/pages/blog/[slug]")
const lazy2 = () => import(".//project/src/pages/contact")
const lazy3 = () => import(".//project/src/pages/index")

// Generated route configuration
export const routes: RouteObject[] = [
  {
    "path": "about",
    "lazy": lazy0
  },
  {
    "path": "blog",
    "children": [
      {
        "path": ":slug",
        "lazy": lazy1
      }
    ]
  },
  {
    "path": "contact",
    "lazy": lazy2
  },
  {
    "path": "",
    "lazy": lazy3
  }
]

export default routes
"
`;

exports[`routeBuilderPluginV2 > Generated Routes Snapshots > should generate correct routes for only sync files > sync-only-routes 1`] = `
"// This file is auto-generated by vite-plugin-route-builder
// Do not edit manually
/* eslint-disable */
// @ts-nocheck

import type { RouteObject } from "react-router"

// Imports for page components
import * as SyncComponent0 from ".//project/src/pages/critical.sync"
import * as SyncComponent1 from ".//project/src/pages/important.sync"
import * as SyncComponent2 from ".//project/src/pages/settings/config.sync"

// Generated route configuration
export const routes: RouteObject[] = [
  {
    "path": "critical",
    "Component": SyncComponent0.Component,
    "loader": SyncComponent0.loader
  },
  {
    "path": "important",
    "Component": SyncComponent1.Component,
    "loader": SyncComponent1.loader
  },
  {
    "path": "settings",
    "children": [
      {
        "path": "config",
        "Component": SyncComponent2.Component,
        "loader": SyncComponent2.loader
      }
    ]
  }
]

export default routes
"
`;

exports[`routeBuilderPluginV2 > Generated Routes Snapshots > should generate correct routes with custom transformPath > custom-transform-routes 1`] = `
"// This file is auto-generated by vite-plugin-route-builder
// Do not edit manually
/* eslint-disable */
// @ts-nocheck

import type { RouteObject } from "react-router"

// Imports for page components


// Generated route configuration
export const routes: RouteObject[] = [
  {
    "path": "",
    "children": [
      {
        "path": "demo"
      },
      {
        "path": "test"
      }
    ]
  }
]

export default routes
"
`;

exports[`routeBuilderPluginV2 > Generated Routes Snapshots > should generate empty routes array when no files found > empty-routes 1`] = `
"// This file is auto-generated by vite-plugin-route-builder
// Do not edit manually
/* eslint-disable */
// @ts-nocheck

import type { RouteObject } from "react-router"

// Imports for page components


// Generated route configuration
export const routes: RouteObject[] = []

export default routes
"
`;

exports[`routeBuilderPluginV2 > should generate correct import statements for sync files > sync-and-async-routes 1`] = `
"// This file is auto-generated by vite-plugin-route-builder
// Do not edit manually
/* eslint-disable */
// @ts-nocheck

import type { RouteObject } from "react-router"

// Imports for page components
import * as SyncComponent0 from ".//project/src/pages/critical.sync"
const lazy0 = () => import(".//project/src/pages/normal")

// Generated route configuration
export const routes: RouteObject[] = [
  {
    "path": "critical",
    "Component": SyncComponent0.Component,
    "loader": SyncComponent0.loader
  },
  {
    "path": "normal",
    "lazy": lazy0
  }
]

export default routes
"
`;
